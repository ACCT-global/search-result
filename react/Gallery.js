import React from 'react'
import classNames from 'classnames'
import { List, WindowScroller, AutoSizer, CellMeasurer, CellMeasurerCache } from 'react-virtualized'
import PropTypes from 'prop-types'
import { Adopt } from 'react-adopt'

import { withRuntimeContext, NoSSR } from 'render'

import { productShape } from './constants/propTypes'
import GalleryItem from './components/GalleryItem'

/**
 * Canonical gallery that displays a list of given products.
 */

const TWO_ITEMS = 2
const ONE_ITEM = 1

/**
 *  Cache to measure the rows height
 * */
const cache = new CellMeasurerCache({
  fixedWidth: true,
  keyMapper: () => 1,
})

const Gallery = ({
  products,
  summary,
  layoutMode,
  runtime: { hints: { mobile } },
  itemWidth,
}) => {

  cache.clearAll()

  const renderItem = (item) => (
    <div
      key={item.productId}
      className="vtex-gallery__item mv2 pa1"
    >
      <GalleryItem
        item={item}
        summary={summary}
        displayMode={layoutMode}
      />
    </div>
  )

  const renderRow = ({ index, key, style, parent, itemsPerRow }) => {
    const from = index * itemsPerRow
    const rowItems = products.slice(from, from + itemsPerRow)

    const containerClasses = classNames('vtex-gallery__row', {
      'vtex-gallery--two-columns': layoutMode === 'small' && mobile,
    })

    return (
      <CellMeasurer
        cache={cache}
        columnIndex={0}
        key={key}
        parent={parent}
        rowIndex={index}
      >
        {({ measure }) => (
          <div className={containerClasses} key={key} style={style} onLoad={measure}>
            {rowItems.map(renderItem)}
          </div>
        )}
      </CellMeasurer>
    )
  }

  const ssrContainer = classNames('vtex-gallery pa3 bn', {
    'vtex-gallery--two-columns': layoutMode == 'small' && mobile
  })

  const ssrFallBack = (
    <div className={ssrContainer}>
      {products.map(renderItem)}
    </div>
  )

  return (
    <NoSSR onSSR={ssrFallBack}>
      <div className="vtex-gallery pa3 bn">
        <Adopt
          key={layoutMode}
          mapper={{
            scroller: <WindowScroller />,
            autoSizer: <AutoSizer disableHeight />,
          }}
          mapProps={({ scroller: { height }, autoSizer: { width } }) => ({ width, height })}
        >
          {({ width, height }) => {
            const itemsPerRow = (layoutMode === 'small' && mobile) ? TWO_ITEMS : (Math.floor(width / itemWidth) || ONE_ITEM)
            const nRows = Math.ceil(products.length / itemsPerRow)

            return (
              <List
                key={layoutMode}
                deferredMeasurementCache={cache}
                autoHeight
                width={width}
                height={height}
                rowCount={nRows}
                rowHeight={cache.rowHeight}
                rowRenderer={args => renderRow({ ...args, itemsPerRow })}
              />
            )
          }}
        </Adopt>
      </div >
    </NoSSR>
  )
}

Gallery.propTypes = {
  /** Products to be displayed. */
  products: PropTypes.arrayOf(productShape),
  /** ProductSummary props. */
  summary: PropTypes.any,
  /** Layout mode of the gallery */
  layoutMode: PropTypes.string,
  /** Item Width. */
  itemWidth: PropTypes.number,
}

Gallery.defaultProps = {
  maxItemsPerPage: 10,
  products: [],
  itemWidth: 300,
}

export default withRuntimeContext(Gallery)
