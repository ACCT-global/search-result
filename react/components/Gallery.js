import React from 'react'
import { PropTypes } from 'prop-types'
import classNames from 'classnames'

import { productShape } from '../constants/propTypes'
import GalleryItem from './GalleryItem'

/**
 * Canonical gallery that displays a list of given products.
 */
const Gallery = props => {
  const { maxItemsPerLine, products, summary } = props

  const className = classNames('vtex-gallery pa3 w-100', {
    'vtex-gallery--max-5': maxItemsPerLine === 5,
    'vtex-gallery--max-4': maxItemsPerLine === 4,
    'vtex-gallery--max-3': maxItemsPerLine === 3,
  })

  return (
    <div className={className}>
      {products.map(item => {
        return (
          <div
            key={item.productId}
            className="vtex-gallery__item mv2 pa1"
          >
            <GalleryItem item={item} summary={summary} />
          </div>
        )
      })}
    </div>
  )
}

Gallery.propTypes = {
  /** Maximum number of items per line. */
  maxItemsPerLine: PropTypes.number.isRequired,
  /** Maximum number of items per page. */
  maxItemsPerPage: PropTypes.number.isRequired,
  /** Products to be displayed. */
  products: PropTypes.arrayOf(productShape),
  /** ProductSummary props. */
  summary: PropTypes.any,
}

Gallery.defaultProps = {
  maxItemsPerLine: 5,
  maxItemsPerPage: 10,
  products: [],
}

export default Gallery
