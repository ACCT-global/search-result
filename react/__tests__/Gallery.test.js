/* eslint-env jest */
import React from 'react'
import { render } from 'test-utils'

import { products, summary } from 'GalleryMocks'

import Gallery from '../Gallery'

describe('<OrderBy />', () => {
  const renderComponent = customProps => {
    const props = {
      products,
      summary,
      ...customProps,
    }

    return render(<Gallery {...props} />)
  }

  it('should match snapshot', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })
})
