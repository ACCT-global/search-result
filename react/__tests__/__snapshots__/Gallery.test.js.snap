// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderBy /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
  />
  <div
    class="gallery flex flex-row flex-wrap items-stretch pa3 bn mh4"
  >
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
    <div
      class="galleryItem"
      style="flex-basis: 20%; max-width: 20%;"
    >
      <div>
         Extension Point 
      </div>
    </div>
  </div>
</DocumentFragment>
`;
